fud2/fud-core/src/exec/plan/egg_planner.rs

```diff
@@ -0,0 +1,260 @@
+    /// For example, if there are 3 states and 2 ops, a term `(root (states x x x) (ops x x))`
+    /// represents the absence of all states and all ops. A term `(root (states 3 x x) (ops x 2))`
+    /// represents a pair of sets, the first containing a single state with `StateRef` 3 and the
+    /// second containing a single op with `OpRef` 2.
+    enum StateLanguage {
+        // The root of a term. This is used to store a pair of "states" and "ops".
+        "root" = Root([Id; 2]),
+        // A list of states.
+        "states" = States(Box<[Id]>),
+        // A list of ops.
+        "ops" = Ops(Box<[Id]>),
+        // Symbolizes the absence of a state or op.
+        "nil" = X,
+        // Symbolizes the presence of a state or op.
+        "one" = C,
+    }
+}
+
+/// Construct an expr for `root` term in `StateLanguage`.
+///
+/// `states` are in included states in the term, and `through` are the included ops. Unspecified
+/// ops and states are maked with "xxx".
+/// `all_states` is an ordered list of all states. `all_ops` is an ordered list of all ops.
+fn language_expr(
+    states: &[StateRef],
+    through: &[OpRef],
+    all_states: &[StateRef],
+    all_ops: &[OpRef],
+) -> RecExpr<StateLanguage> {
+    let mut expr: RecExpr<StateLanguage> = Default::default();
+    // Collect states into an expr.
+    let state_ids = all_states
+        .iter()
+        .map(|s| {
+            if states.contains(s) {
+                expr.add(StateLanguage::C)
+            } else {
+                expr.add(StateLanguage::X)
+            }
+        })
+        .collect::<Vec<_>>();
+    let state_expr =
+        expr.add(StateLanguage::States(state_ids.into_boxed_slice()));
+
+    // Collect ops into an expr.
+    let op_ids = all_ops
+        .iter()
+        .map(|o| {
+            if through.contains(o) {
+                expr.add(StateLanguage::C)
+            } else {
+                expr.add(StateLanguage::X)
+            }
+        })
+        .collect::<Vec<_>>();
+    let op_expr = expr.add(StateLanguage::Ops(op_ids.into_boxed_slice()));
+
+    expr.add(StateLanguage::Root([state_expr, op_expr]));
+    expr
+}
+
+/// Creates a `Rewrite` corresponding to `op` and `op_ref` for `StateLanguage`.
+///
+/// `through` is all of the ops required to be used by a plan the planner is attempting to find.
+/// This is used as an optimization so only the ops specified in `through` are needed to be kept
+/// track of in `ops` terms.
+///
+/// `all_states` is an ordered list of all states. `all_ops` is an ordered list of all ops.
+///
+/// This rewrite will look like `(root (states <state list>) (ops <op list>)) =>
+/// (root (states <state list - op inputs + op outputs>) (ops <op list + op>))`
+///
+/// As an example, given 3 ops and 4 states, if op 2 takes states 2 and 3 to state 4, the rewrite
+/// would take
+/// `(root (states 1 2 3 xxx) (ops 1 xxx xxx)) => (root (states 1 xxx xxx r) (ops 1 2 xxx)`.
```

Comment:
Notice some small typos. Do you mean `4` instead of `r`? The rhs of the rewrite is missing a paren.